python_modules/dagster/dagster/_utils/caching_instance_queryer.py

```diff
@@ -150,8 +185,11 @@ def get_latest_materialization_record(
         else:
             asset_partition = asset
 
-        # fancy caching only applies to after_cursor
         if before_cursor is not None:
+            latest_record = self._latest_materialization_record_cache.get(asset_partition)
```

Comment:
If the before_cursor is greater than the storage id of the latest record for an asset, then we know that the most recent record before that cursor is the same as the latest one.